### PR TITLE
Run Babel compiler lazily, in the ImportScanner.

### PR DESCRIPTION
**Pull request by benjamn**
One of the weaknesses of Meteor's compiler plugin system is that compiler plugins run before the `ImportScanner` has a chance to identify which modules are actually needed, so unused files may be compiled unnecessarily.

Historically, the reason for this behavior was that the `processFilesForTarget` method of compiler plugins expects to receive a complete list of files to be compiled, rather than compiling files individually.

More recently, `BabelCompiler` and other plugins have begun supporting a [`processOneFileForTarget`](https://github.com/meteor/meteor/blob/5b39d4b2da3cda2137db7cc68b595ba27ca2c214/packages/babel-compiler/babel-compiler.js#L36) helper method, which can be called by other plugins, [like `coffeescript`](https://github.com/meteor/meteor/blob/5b39d4b2da3cda2137db7cc68b595ba27ca2c214/packages/non-core/coffeescript/plugin/compile-coffeescript.js#L100).

This new method is important, because it signals the compiler plugin is capable of compiling files individually, and does not somehow depend on getting to see a complete list of files at one time. This flexibility enables us to defer compilation of files until the `ImportScanner` first accesses the `data` property of the file, and this PR ensures the `ImportScanner` will only access that property when it scans the file as a dependency.

This is not the implementation I would have used if I were writing this code from scratch (property getters are gross), but it's a reasonable tweak to the existing system, and it thankfully does not require refactoring all of our code that handles JS resources so that they have a `getData` method instead of a `data` property.

### CODE REVIEW

**Comment by benjamn** [comment]
Note that this optimization benefits any compiler plugin that defines a `processOneFileForTarget` method, not just the `BabelCompiler`, as the title of the PR might suggest.

File: tools/isobuild/compiler-plugin.js [hunk 1]
```diff
@@ -179,17 +179,35 @@ export class CompilerPluginProcessor {
...
+          if (typeof processOne === "function") {
+            // Instead of running the processOneFileForTarget method now,
+            // call the addLazyJavaScript method of the InputFile so that
+            // compilation can be performed when/if necessary. Compiler
+            // plugins can opt into this behavior by defining a method
+            // called processOneFileForTarget, which indicates they are
+            // willing to compile individual files, instead of a complete
+            // list of files all at once (processFilesForTarget).
+            resourceSlots.forEach(slot => {
+              new InputFile(slot).addLazyJavaScript(buildmessage.markBoundary(
```
Comment (benjamn) [inline tools/isobuild/compiler-plugin.js]: Concern: we don't actually know at this point that the compiler plugin would have called `addJavaScript` with the result of `processOneFileForTarget`, unless we assume that `processOneFileForTarget` is only used for JS files.

**Comment by benjamn** [comment]
Those test failures are probably legitimate. I'll keep investigating this.

**Comment by glasser** [comment]
Can you clarify what you're looking for in a review from me? I assume you tagged me because I wrote some of this code originally, but I definitely have a weak grasp on how this stuff works these days.

**Comment by glasser** [comment]
(sorry, hit wrong button)

**Comment by benjamn** [comment]
@glasser I don't need a full review. I was mostly wondering if you recall the motivation for `processFilesForTarget` receiving all files at once, and if it seems reasonable to allow compiler plugins to compile individual files one at a time (in an opt-in way). It's possible that too much has changed since you last thought about this, but I'm curious if it rings any bells.

**Comment by glasser** [comment]
Here's the original design doc: https://meteor.quip.com/YLcCAxtjuwsL

The short answer is that the original compiler plugin API only allowed a single file to be processed at a time, and also it entirely ran at package build time, which meant that you couldn't do anything where the way you compiled a package (to be published to Atmosphere, eg) depended on any other package or your app.  My solution was to make a new compiler plugin API that ran at app build time, and received all of the files.  This allowed plugins to do more, but isn't as streamy as it could be (though some other aspects of this solution enabled plugins to do their own caching in a way that couldn't be done before).

This was before there was such thing as an import scanner; I guess my theory was that plugins would basically have their own import scanners.

You might as well read the design doc (for historical interest if nothing else) but I'd be shocked if whatever you're proposing isn't a good idea :)

**Comment by hwillson** [comment]
> Those test failures are probably legitimate. I'll keep investigating this.

@benjamn, just a quick observation I noticed while reviewing this for our meeting today. It looks like the failing tests are string matching against error stack traces, and the stack traces have changed with these modifications. Looking at the failing `source maps from an app` test for example, the new error stack trace looks like:

```
Error: Should be line 3!
  at app/app.js:11:11
  ...
```

but the self-test is expecting:

```
Error: Should be line 3!
  at throw.js:3:x
  ...
```

Interestingly, if we remove `ecmascript` from the `app-throws-error` test app that's being used, the resulting stack trace looks like:

```
Error: Should be line 3!
  at app/throw.js:3:11
```

So it's closer, but still fails the more rigid `run.match(matchPathRegexp('at throw\\.js:3\\b'));` check the test has in place (and that worked before these changes).

**Comment by benjamn** [comment]
I need to spend some more time thinking about how this optimization can be less of a hack. In particular, the `processOneFileForTarget` is not appropriate for compiler plugins that call `intputFile.addJavaScript` multiple (or even zero) times, or call other methods like `inputFile.addStylesheet` (as there is no `inputFile.addLazyStylesheet` to match `inputFile.addLazyJavaScript`).

The test failures also tell me that triggering compilation lazily from a `.data` getter property is not as negative-consequence-free as I hoped, presumably because compiler plugins can have side effects that are observed before `resource.data` is first accessed.

A more thorough solution to this problem would be to overhaul the compiler plugin system so that files can be compiled on-demand, when encountered by the `ImportScanner`. There's a good chance that won't be much harder than getting this hack to work correctly, in which case we should just fix the problem properly.

**Pull request closed without merging**